sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/restriction/PartitionRestrictionTracker.java

```diff
@@ -0,0 +1,171 @@
+    this.progressChecker = progressChecker;
+    this.timeSupplier = () -> Timestamp.now();
+  }
+
+  @VisibleForTesting
+  public void setTimeSupplier(Supplier<Timestamp> timeSupplier) {
+    this.timeSupplier = timeSupplier;
+  }
+
+  @Override
+  public @Nullable SplitResult<PartitionRestriction> trySplit(double fractionOfRemainder) {
+    final SplitResult<PartitionRestriction> splitResult =
+        splitter.trySplit(fractionOfRemainder, lastClaimedPosition, restriction);
+    if (splitResult != null) {
+      PartitionRestriction restrictionFromSplit =
+          Optional.ofNullable(splitResult.getPrimary()).orElse(this.restriction);
+      this.restriction = restrictionFromSplit;
+    }
+    return splitResult;
+  }
+
+  @Override
+  public boolean tryClaim(PartitionPosition position) {
+    final boolean canClaim = claimer.tryClaim(restriction, lastClaimedPosition, position);
+
+    if (canClaim) {
+      this.lastClaimedPosition = position;
+    }
+
+    return canClaim;
+  }
+
+  @Override
+  public Progress getProgress() {
+    final String token =
+        Optional.ofNullable(restriction.getMetadata())
+            .map(PartitionRestrictionMetadata::getPartitionToken)
+            .orElse(null);
+    BigDecimal end;
+    if (restriction.getEndTimestamp().compareTo(Timestamp.MAX_VALUE) == 0) {
+      // When the given end timestamp equals to Timestamp.MAX_VALUE, this means that
+      // the end timestamp is not specified which should be a streaming job. So we
+      // use now() as the end timestamp.
+      end = BigDecimal.valueOf(timeSupplier.get().getSeconds());
+    } else {
+      end = BigDecimal.valueOf(restriction.getEndTimestamp().getSeconds());
+    }
+    final Progress progress = progressChecker.getProgress(restriction, lastClaimedPosition, end);
+    LOG.debug("[" + token + "] Progress is " + progress);
+    return progress;
+  }
+
+  @Override
+  public PartitionRestriction currentRestriction() {
+    return restriction;
+  }
+
+  @Override
+  public void checkDone() throws IllegalStateException {
+    if (restriction.getMode() == STOP) {
+      return;
+    }
+    if (lastClaimedPosition != null) {
+      final PartitionMode currentMode = lastClaimedPosition.getMode();
+      if (currentMode == QUERY_CHANGE_STREAM) {
+        // If the mode is QUERY_CHANGE_STREAM, and the end of the timestamp range wasn't attempted,
+        // throw an exception
+        final Timestamp nextPosition = next(lastClaimedPosition.getTimestamp().get());
+        if (nextPosition.compareTo(restriction.getEndTimestamp()) < 0) {
+          throw new IllegalStateException(
+              String.format(
+                  "Last attempted key was %s in range %s, claiming work in [%s, %s) was not"
+                      + " attempted",
+                  lastClaimedPosition.getTimestamp().get(),
+                  restriction,
+                  nextPosition,
+                  restriction.getEndTimestamp()));
+        }
```

Comment:
Do I understand it correctly that for the modes != {STOP, QUERY_CHANGE_STREAM, DONE} we will always succeed if we claimed a position?

Reply:
No....


For modes STOP/DONE, they should always succeed if we claimed a position

For mode QUERY_CHANGE_STREAM, it should only succeed if we claimed endTimestamp - 1, for restriction [startTimestamp, endTimestamp)

For all other modes, an error should be thrown, since in the splitter, I don't believe the primary restriction will ever be one of the other modes after a split.

Reply:
Ah ok, so when we split and we are currently in the WAIT_FOR_CHILD_PARTITIONS or in the UPDATE_STATE modes, we will update the primary restriction to be STOP, which will make the checkDone call succeed, right?

Reply:
Yes, exactly!